test.lua

```diff
@@ -16,6 +17,54 @@ strict.on()
 --                          local support functions                          --
 -- ========================================================================= --
 
+-- Strip pcall()'s true or re-raise catched error.
+local function wait_cond_finish(res)
+    -- res: 1: pcall_ok
+    --      2: pcall_error / cond_function_res_1
+    --      3: cond_function_res_2
+    --      4: cond_function_res_3
+    --      ...
+    --
+    -- cond_function_res_1 is the 'ok' / 'not ok' indicator for
+    -- wait_cond(). But we return it: this is deliberate choice.
+    -- This way a user may understand, whether we finally hit
+    -- timeout or not.
+    if not res[1] then
+        error(res[2])
+    end
+    return select(2, unpack(res))
+end
```

Comment:
I guess it should be done in the way as box.atomic work (https://github.com/tarantool/tarantool/blob/298eebf7c8e8dc2c985f20e22d35ef5c1bfd7c17/src/box/lua/schema.lua#L356)

Reply:
If I get it right you want a rollback in case of cond function fails.
However everywhere wait_cond() is used for polling stats from expirationd and wait_cond() doesn't execute any operations with data. So using something like `box.atomic()` is not applicable here.